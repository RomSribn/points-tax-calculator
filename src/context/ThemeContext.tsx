import React from 'react';
import { TTheme } from '@utils/interfaces';

type TodoContextType = {
  theme: TTheme;
  updateTheme?: () => void;
};

const defaultState: { theme: TTheme } = {
  theme: 'dark'
};

const ThemeContext: React.Context<TodoContextType> = React.createContext(defaultState);
const ThemeProvider = ThemeContext.Provider;
const ThemeConsumer = ThemeContext.Consumer;

export { ThemeContext, ThemeProvider, ThemeConsumer };
