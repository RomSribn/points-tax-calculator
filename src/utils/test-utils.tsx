import React, { ReactElement } from 'react';
import { render, RenderOptions } from '@testing-library/react';
import { ThemeProvider, defaultState } from '@context/ThemeContext';
import '@testing-library/jest-dom';

const AppProvider = ({ children }: { children: React.ReactNode }) => {
  return <ThemeProvider value={defaultState}>{children}</ThemeProvider>;
};

const customRender = (ui: ReactElement, options?: Omit<RenderOptions, 'wrapper'>) =>
  render(ui, { wrapper: AppProvider, ...options });

export * from '@testing-library/react';
export { customRender };
