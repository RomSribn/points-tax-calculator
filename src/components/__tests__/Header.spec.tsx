import React from 'react';
import userEvent from '@testing-library/user-event';
import { customRender, screen } from '@utils/test-utils';
import { Header } from '@components/index';
import { ThemeProvider } from '@context/ThemeContext';

describe('Header', () => {
  test('Should renders the component correctly', () => {
    const wrapper = customRender(<Header />);

    // Assert that the header is rendered
    expect(screen.getByRole('banner')).not.toBeNull();
    expect(screen.getByRole('button', { name: 'Toggle navigation' })).not.toBeNull();
    expect(screen.getByRole('navigation')).not.toBeNull();

    expect(wrapper).toMatchSnapshot();
  });

  test('Should toggles the theme when the button is clicked', () => {
    const updateThemeMock = jest.fn();
    const theme = 'light';

    customRender(
      <ThemeProvider value={{ theme, updateTheme: updateThemeMock }}>
        <Header />
      </ThemeProvider>
    );

    // Mock the updateTheme function from the ThemeContext
    const useContextSpy = jest.spyOn(React, 'useContext');
    useContextSpy.mockReturnValue({ theme, updateTheme: updateThemeMock });

    const button = screen.getByTestId('darkTheme');

    // Assert that the initial theme is 'dark'
    expect(button).not.toBeNull();
    expect(screen.getByTestId('moon-icon')).not.toBeNull();

    // Simulate clicking the button to toggle the theme
    userEvent.click(button);

    // Assert that the updateTheme function is called
    expect(updateThemeMock).toHaveBeenCalled();
  });
});
