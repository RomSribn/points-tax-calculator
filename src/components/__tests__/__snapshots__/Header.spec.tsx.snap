// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header Should renders the component correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <header
        class="light-bb"
      >
        <nav
          class="navbar navbar-expand-lg navbar-light"
        >
          <button
            aria-controls="basic-navbar-nav"
            aria-label="Toggle navigation"
            class="navbar-toggler collapsed"
            type="button"
          >
            <span
              class="navbar-toggler-icon"
            />
          </button>
          <div
            class="navbar-collapse collapse"
            id="basic-navbar-nav"
          >
            <div
              class="navbar-nav ml-auto navbar-nav"
            >
              <div
                class="header-custom-icon dropdown"
              >
                <button
                  class="btn btn-default"
                  data-testid="darkTheme"
                  type="button"
                >
                  <img
                    data-testid="sunny-icon"
                    src="/img/icon/sunny.svg"
                  />
                </button>
              </div>
            </div>
          </div>
        </nav>
      </header>
    </div>
  </body>,
  "container": <div>
    <header
      class="light-bb"
    >
      <nav
        class="navbar navbar-expand-lg navbar-light"
      >
        <button
          aria-controls="basic-navbar-nav"
          aria-label="Toggle navigation"
          class="navbar-toggler collapsed"
          type="button"
        >
          <span
            class="navbar-toggler-icon"
          />
        </button>
        <div
          class="navbar-collapse collapse"
          id="basic-navbar-nav"
        >
          <div
            class="navbar-nav ml-auto navbar-nav"
          >
            <div
              class="header-custom-icon dropdown"
            >
              <button
                class="btn btn-default"
                data-testid="darkTheme"
                type="button"
              >
                <img
                  data-testid="sunny-icon"
                  src="/img/icon/sunny.svg"
                />
              </button>
            </div>
          </div>
        </div>
      </nav>
    </header>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
